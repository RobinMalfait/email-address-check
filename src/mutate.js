function mutate(mutator) {
  return (...input) => mutator(...input);
}

export default mutate;
