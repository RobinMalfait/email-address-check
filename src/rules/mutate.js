export default mutator => (username, hostname) => {
  const mutation = mutator(username, hostname);

  if (Array.isArray(mutation)) {
    const [new_username, new_hostname] = mutation;
    return {
      username: new_username || username,
      hostname: new_hostname || hostname
    };
  }

  if (typeof mutation === "object") {
    const { username, hostname } = mutation;
    return { username, hostname };
  }

  if (typeof mutation === "string") {
    return {
      username: mutation,
      hostname
    };
  }

  return {
    username,
    hostname
  };
};
