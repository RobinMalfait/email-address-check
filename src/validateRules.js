function validateRules(rules) {
  return async input => {
    for (let i = 0; i < rules.length; i++) {
      const value = await rules[i](input);

      // Early return
      if (value === true) {
        return true;
      }

      // It's a mutator!
      if (value !== undefined) {
        input = value;
      }
    }
  };
}

export default validateRules;
