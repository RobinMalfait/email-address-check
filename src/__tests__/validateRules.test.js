import validateRules from "../validateRules";
import mutate from "../mutate";

describe("validateRules", () => {
  it("should early return when a rule returns true", () => {
    let count = 0;
    const rules = [
      () => {
        count++;
        return;
      },
      () => {
        count++;
        return true;
      },
      () => {
        count++;
        return;
      }
    ];

    return validateRules(rules)("some input").then(result => {
      expect(count).toEqual(2);
      expect(result).toMatchSnapshot();
    });
  });

  it("should mutate the input if a mutator is defined", () => {
    const rules = [
      mutate(input => "my new input"),
      new_input => {
        throw new Error(new_input);
      }
    ];

    return expect(validateRules(rules)("some input")).rejects.toMatchSnapshot();
  });
});
