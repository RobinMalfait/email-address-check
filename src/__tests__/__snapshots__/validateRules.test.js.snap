// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateRules should early return when a rule returns true 1`] = `true`;

exports[`validateRules should mutate the input if a mutator is defined 1`] = `[Error: my new input]`;
